rpc/client/eventstream/eventstream.go

```diff
@@ -0,0 +1,188 @@
+// Package eventstream implements a convenience client for the Events method
+// of the Tendermint RPC service, allowing clients to observe a resumable
+// stream of events matching a query.
+package eventstream
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/tendermint/tendermint/rpc/coretypes"
+)
+
+// Client is the subset of the RPC client interface consumed by Stream.
+type Client interface {
+	Events(ctx context.Context, req *coretypes.RequestEvents) (*coretypes.ResultEvents, error)
+}
+
+// ErrStopRunning is returned by a Run callback to signal that no more events
+// are wanted and that Run should return.
+var ErrStopRunning = errors.New("stop accepting events")
+
+// A Stream cpatures the state of a streaming event subscription.
+type Stream struct {
+	filter     *coretypes.EventFilter // the query being streamed
+	batchSize  int                    // request batch size
+	newestSeen string                 // from the latest item matching our query
+	waitTime   time.Duration          // the long-polling interval
+	client     Client
+}
+
+// New constructs a new stream for the given query and options.
+// If opts == nil, the stream uses default values as described by
+// StreamOptions.  This function will panic if cli == nil.
+func New(cli Client, query string, opts *StreamOptions) *Stream {
+	if cli == nil {
+		panic("eventstream: nil client")
+	}
+	return &Stream{
+		filter:     &coretypes.EventFilter{Query: query},
+		batchSize:  opts.batchSize(),
+		newestSeen: opts.resumeFrom(),
+		waitTime:   opts.waitTime(),
+		client:     cli,
+	}
+}
+
+// Run polls the service for events matching the query, and calls accept for
+// each such event. Run handles pagination transparently, and delivers events
+// to accept in order of publication.
+//
+// Run continues until ctx ends or accept reports an error.  If accept returns
+// ErrStopRunning, Run returns nil; otherwise Run returns the error reported by
+// accept or ctx.  Run also returns an error if the server reports an error
+// from the Events method.
+//
+// If the stream falls behind the event log on the server, Run will stop and
+// report an error of concrete type *MissedItemsError.  Call Reset to reset the
+// stream to the head of the log, and call Run again to resume.
+func (s *Stream) Run(ctx context.Context, accept func(*coretypes.EventItem) error) error {
+	after := s.newestSeen // request (only) items after this
+	var pageCursor string // if non-empty, page through items before this
+	for {
+		var items []*coretypes.EventItem
+
+		// Fetch another batch of events from the service.
+		for {
+			rsp, err := s.client.Events(ctx, &coretypes.RequestEvents{
```

Comment:
I had to read this all a bunch of times to convince myself that we weren't throwing away the items list in between iterations. 

My inclination is:
- do some producer/consumer thing with processing events so you smooth out the impact on consumers rather than being really bursty
- factor some of the pagination handling into other fuctions so the `Run` logic is easier to read. 

Having said all that, once I got my brain around this, I like it a lot.

Reply:
Re: producer/consumer. I really don't want to introduce concurrency here without a good reason. I think the impact on the consumer shouldn't matter, since we're not pushing back anything—if the receiver is slow, we just won't fetch more data. (Of course, if they're slow enough, we could fall off the eventlog window—but that is a problem even if we flatten out the profile)

Re: Factor. Let me see what I can do with that.

Reply:
PTAL, I pulled out the inner pagination loop into a separate method.